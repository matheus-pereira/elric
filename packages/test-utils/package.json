{
  "name": "@elric/test-utils",
  "version": "27.4.6",
  "private": true,
  "license": "MIT",
  "main": "./build/index.js",
  "types": "./build/index.d.ts",
  "exports": {
    ".": {
      "types": "./build/index.d.ts",
      "default": "./build/index.js"
    },
    "./package.json": "./package.json"
  },
  "dependencies": {
    "@elric/types": "^27.4.2",
    "@types/node": "*",
    "ansi-regex": "^5.0.1",
    "ansi-styles": "^5.0.0",
    "pretty-format": "^27.4.6",
    "semver": "^7.3.2"
  },
  "devDependencies": {
    "@types/semver": "^7.1.0"
  },
  "engines": {
    "node": "^10.13.0 || ^12.13.0 || ^14.15.0 || >=15.0.0"
  }
}
