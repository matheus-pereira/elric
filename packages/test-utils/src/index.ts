/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export {alignedAnsiStyleSerializer} from './alignedAnsiStyleSerializer';

export {
  iselricJasmineRun,
  skipSuiteOnJasmine,
  skipSuiteOnelricCircus,
  onNodeVersions,
} from './ConditionalTest';

export {makeGlobalConfig, makeProjectConfig} from './config';
