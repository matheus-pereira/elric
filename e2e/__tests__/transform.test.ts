/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {tmpdir} from 'os';
import * as path from 'path';
import * as fs from 'graceful-fs';
import {wrap} from 'elric-snapshot-serializer-raw';
import {onNodeVersions} from '@elric/test-utils';
import {
  cleanup,
  copyDir,
  createEmptyPackage,
  extractSummary,
  linkelricPackage,
  runYarnInstall,
} from '../Utils';
import runelric, {json as runWithJson} from '../runelric';

describe('babel-elric', () => {
  const dir = path.resolve(__dirname, '..', 'transform/babel-elric');

  beforeEach(() => {
    runYarnInstall(dir);
  });

  it('runs transpiled code', () => {
    // --no-cache because babel can cache stuff and result in false green
    const {json} = runWithJson(dir, ['--no-cache']);
    expect(json.success).toBe(true);
    expect(json.numTotalTests).toBeGreaterThanOrEqual(2);
  });

  it('instruments only specific files and collects coverage', () => {
    const {stdout} = runelric(dir, ['--coverage', '--no-cache'], {
      stripAnsi: true,
    });
    expect(stdout).toMatch('covered.js');
    expect(stdout).not.toMatch('notCovered.js');
    expect(stdout).not.toMatch('excludedFromCoverage.js');
    // coverage result should not change
    expect(wrap(stdout)).toMatchSnapshot();
  });
});

describe('babel-elric ignored', () => {
  const dir = path.resolve(__dirname, '..', 'transform/babel-elric-ignored');

  it('tells user to match ignored files', () => {
    // --no-cache because babel can cache stuff and result in false green
    const {exitCode, stderr} = runelric(dir, ['--no-cache']);
    expect(exitCode).toBe(1);
    expect(wrap(extractSummary(stderr).rest)).toMatchSnapshot();
  });
});

describe('babel-elric with manual transformer', () => {
  const dir = path.resolve(__dirname, '..', 'transform/babel-elric-manual');

  beforeEach(() => {
    runYarnInstall(dir);
  });

  it('runs transpiled code', () => {
    // --no-cache because babel can cache stuff and result in false green
    const {json} = runWithJson(dir, ['--no-cache']);
    expect(json.success).toBe(true);
    expect(json.numTotalTests).toBeGreaterThanOrEqual(1);
  });
});

// babel-elric is automatically linked at the root because it is a workspace now
// a way to test this in isolation is to move the test suite into a temp folder
describe('no babel-elric', () => {
  const dir = path.resolve(__dirname, '..', 'transform/no-babel-elric');
  // doing test in a temp directory because we don't want elric node_modules affect it
  const tempDir = path.resolve(tmpdir(), 'transform-no-babel-elric');

  beforeEach(() => {
    cleanup(tempDir);
    createEmptyPackage(tempDir);
    copyDir(dir, tempDir);
    linkelricPackage('babel-elric', tempDir);
  });

  test('fails with syntax error on flow types', () => {
    const {stderr} = runelric(tempDir, ['--no-cache', '--no-watchman']);
    expect(stderr).toMatch(/FAIL.*failsWithSyntaxError/);
    expect(stderr).toMatch('Unexpected token');
  });

  test('instrumentation with no babel-elric', () => {
    const {stdout} = runelric(
      tempDir,
      ['--no-cache', '--coverage', '--no-watchman'],
      {stripAnsi: true},
    );
    expect(stdout).toMatch('covered.js');
    expect(stdout).not.toMatch('excludedFromCoverage.js');
    // coverage result should not change
    expect(wrap(stdout)).toMatchSnapshot();
  });
});

describe('custom transformer', () => {
  const dir = path.resolve(
    __dirname,
    '..',
    'transform/custom-instrumenting-preprocessor',
  );

  it('proprocesses files', () => {
    const {json, stderr} = runWithJson(dir, ['--no-cache']);
    expect(stderr).toMatch(/FAIL/);
    expect(stderr).toMatch(/instruments by setting.*global\.__INSTRUMENTED__/);
    expect(json.numTotalTests).toBe(2);
    expect(json.numPassedTests).toBe(1);
    expect(json.numFailedTests).toBe(1);
  });

  it('instruments files', () => {
    const {stdout, exitCode} = runelric(dir, ['--no-cache', '--coverage'], {
      stripAnsi: true,
    });
    // coverage should be empty because there's no real instrumentation
    expect(wrap(stdout)).toMatchSnapshot();
    expect(exitCode).toBe(0);
  });
});

describe('multiple-transformers', () => {
  const dir = path.resolve(__dirname, '..', 'transform/multiple-transformers');

  beforeEach(() => {
    runYarnInstall(dir);
  });

  it('transforms dependencies using specific transformers', () => {
    const {json, stderr} = runWithJson(dir, ['--no-cache']);

    expect(stderr).toMatch(/PASS/);
    expect(json.numTotalTests).toBe(1);
    expect(json.numPassedTests).toBe(1);
  });
});

describe('ecmascript-modules-support', () => {
  const dir = path.resolve(
    __dirname,
    '..',
    'transform/ecmascript-modules-support',
  );

  it('runs transpiled code', () => {
    // --no-cache because babel can cache stuff and result in false green
    const {json} = runWithJson(dir, ['--no-cache']);
    expect(json.success).toBe(true);
    expect(json.numTotalTests).toBeGreaterThanOrEqual(1);
  });
});

describe('transformer-config', () => {
  const dir = path.resolve(__dirname, '..', 'transform/transformer-config');

  beforeEach(() => {
    runYarnInstall(dir);
  });

  it('runs transpiled code', () => {
    // --no-cache because babel can cache stuff and result in false green
    const {json} = runWithJson(dir, ['--no-cache']);
    expect(json.success).toBe(true);
    expect(json.numTotalTests).toBeGreaterThanOrEqual(1);
  });

  it('instruments only specific files and collects coverage', () => {
    const {stdout} = runelric(dir, ['--coverage', '--no-cache'], {
      stripAnsi: true,
    });
    expect(stdout).toMatch('Covered.js');
    expect(stdout).not.toMatch('NotCovered.js');
    expect(stdout).not.toMatch('ExcludedFromCoverage.js');
    // coverage result should not change
    expect(wrap(stdout)).toMatchSnapshot();
  });
});

describe('transformer caching', () => {
  const dir = path.resolve(__dirname, '../transform/cache');
  const transformedFile = path.resolve(dir, './common-file.js');

  it('does not rerun transform within worker', () => {
    // --no-cache because babel can cache stuff and result in false green
    const {stdout} = runelric(dir, ['--no-cache', '-w=2']);

    const loggedFiles = stdout.split('\n');

    // Verify any lines logged are _just_ the file we care about
    loggedFiles.forEach(line => {
      expect(line).toBe(transformedFile);
    });

    // We run with 2 workers, so the file should be transformed twice
    expect(loggedFiles).toHaveLength(2);
  });
});

describe('transform-snapshotResolver', () => {
  const dir = path.resolve(
    __dirname,
    '..',
    'transform/transform-snapshotResolver',
  );
  const snapshotDir = path.resolve(dir, '__snapshots__');
  const snapshotFile = path.resolve(snapshotDir, 'snapshot.test.js.snap');

  const cleanupTest = () => {
    if (fs.existsSync(snapshotFile)) {
      fs.unlinkSync(snapshotFile);
    }
    if (fs.existsSync(snapshotDir)) {
      fs.rmdirSync(snapshotDir);
    }
  };

  beforeAll(() => {
    runYarnInstall(dir);
  });
  beforeEach(cleanupTest);
  afterAll(cleanupTest);

  it('should transform the snapshotResolver', () => {
    const result = runelric(dir, ['-w=1', '--no-cache', '--ci=false']);

    expect(result.stderr).toMatch('1 snapshot written from 1 test suite');

    const contents = require(snapshotFile);
    expect(contents).toHaveProperty(
      'snapshots are written to custom location 1',
    );
  });
});

describe('transform-environment', () => {
  const dir = path.resolve(__dirname, '../transform/transform-environment');

  it('should transform the environment', () => {
    const {json, stderr} = runWithJson(dir, ['--no-cache']);
    expect(stderr).toMatch(/PASS/);
    expect(json.success).toBe(true);
    expect(json.numPassedTests).toBe(1);
  });
});

describe('transform-runner', () => {
  const dir = path.resolve(__dirname, '../transform/transform-runner');

  it('should transform runner', () => {
    const {json, stderr} = runWithJson(dir, ['--no-cache']);
    expect(stderr).toMatch(/PASS/);
    expect(json.success).toBe(true);
    expect(json.numPassedTests).toBe(1);
  });
});

describe('transform-testrunner', () => {
  const dir = path.resolve(__dirname, '../transform/transform-testrunner');

  it('should transform testRunner', () => {
    const {json, stderr} = runWithJson(dir, ['--no-cache']);
    expect(stderr).toMatch(/PASS/);
    expect(json.success).toBe(true);
    expect(json.numPassedTests).toBe(1);
  });
});

onNodeVersions('>=12.17.0', () => {
  describe('esm-transformer', () => {
    const dir = path.resolve(__dirname, '../transform/esm-transformer');

    it('should transform with transformer written in ESM', () => {
      const {json, stderr} = runWithJson(dir, ['--no-cache']);
      expect(stderr).toMatch(/PASS/);
      expect(json.success).toBe(true);
      expect(json.numPassedTests).toBe(1);
    });
  });

  describe('async-transformer', () => {
    const dir = path.resolve(__dirname, '../transform/async-transformer');

    it('should transform with transformer with only async transforms', () => {
      const {json, stderr} = runWithJson(dir, ['--no-cache'], {
        nodeOptions: '--experimental-vm-modules --no-warnings',
      });
      expect(stderr).toMatch(/PASS/);
      expect(json.success).toBe(true);
      expect(json.numPassedTests).toBe(2);
    });
  });

  describe('babel-elric-async', () => {
    const dir = path.resolve(__dirname, '../transform/babel-elric-async');

    beforeAll(() => {
      runYarnInstall(dir);
    });

    it("should use babel-elric's async transforms", () => {
      const {json, stderr} = runWithJson(dir, ['--no-cache'], {
        nodeOptions: '--experimental-vm-modules --no-warnings',
      });
      expect(stderr).toMatch(/PASS/);
      expect(json.success).toBe(true);
      expect(json.numPassedTests).toBe(1);
    });
  });

  describe('transform-esm-runner', () => {
    const dir = path.resolve(__dirname, '../transform/transform-esm-runner');
    test('runs test with native ESM', () => {
      const {json, stderr} = runWithJson(dir, ['--no-cache'], {
        nodeOptions: '--experimental-vm-modules --no-warnings',
      });

      expect(stderr).toMatch(/PASS/);
      expect(json.success).toBe(true);
      expect(json.numPassedTests).toBe(1);
    });
  });

  describe('transform-esm-testrunner', () => {
    const dir = path.resolve(
      __dirname,
      '../transform/transform-esm-testrunner',
    );
    test('runs test with native ESM', () => {
      const {json, stderr} = runWithJson(dir, ['--no-cache'], {
        nodeOptions: '--experimental-vm-modules --no-warnings',
      });

      expect(stderr).toMatch(/PASS/);
      expect(json.success).toBe(true);
      expect(json.numPassedTests).toBe(1);
    });
  });
});
