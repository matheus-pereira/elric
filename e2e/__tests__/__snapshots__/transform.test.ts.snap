// elric Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-elric ignored tells user to match ignored files 1`] = `
FAIL __tests__/ignoredFile.test.js
  ‚óè Test suite failed to run

    babel-elric: Babel ignores __tests__/ignoredFile.test.js - make sure to include the file in elric's transformIgnorePatterns as well.

      at assertLoadedBabelConfig (../../../packages/babel-elric/build/index.js:130:11)
`;

exports[`babel-elric instruments only specific files and collects coverage 1`] = `
------------|---------|----------|---------|---------|-------------------
File        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
------------|---------|----------|---------|---------|-------------------
All files   |   83.33 |      100 |      50 |      80 |                   
 covered.js |   83.33 |      100 |      50 |      80 | 13                
------------|---------|----------|---------|---------|-------------------
`;

exports[`custom transformer instruments files 1`] = `
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |       0 |        0 |       0 |       0 |                   
----------|---------|----------|---------|---------|-------------------
`;

exports[`no babel-elric instrumentation with no babel-elric 1`] = `
------------|---------|----------|---------|---------|-------------------
File        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
------------|---------|----------|---------|---------|-------------------
All files   |   83.33 |      100 |      50 |      80 |                   
 covered.js |   83.33 |      100 |      50 |      80 | 13                
------------|---------|----------|---------|---------|-------------------
`;

exports[`transformer-config instruments only specific files and collects coverage 1`] = `
------------|---------|----------|---------|---------|-------------------
File        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
------------|---------|----------|---------|---------|-------------------
All files   |   83.33 |      100 |      50 |      80 |                   
 Covered.js |   83.33 |      100 |      50 |      80 | 13                
------------|---------|----------|---------|---------|-------------------
`;
